File: src/main/java/com/projectswg/holocore/services/gameplay/training/AvailableSkillsWindow.java
package com.projectswg.holocore.services.gameplay.training;

import com.projectswg.holocore.resources.support.global.player.Player;
import com.projectswg.holocore.resources.support.global.zone.sui.SuiListBox;
import com.projectswg.holocore.resources.support.global.zone.sui.SuiWindow;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public class AvailableSkillsWindow {
	
	private static final String GREEN = "\\#00FF00";
	private static final String RED = "\\#FF0000";
	private static final String WHITE = "\\#FFFFFF";
	
	public void show(String professionName, Player player) {
		Profession profession = new Profession(professionName);
		SkillsSdbSkillRepository skillRepository = new SkillsSdbSkillRepository();
		Trainee trainee = new TraineeImpl(player, skillRepository);
		Training training = new Training(skillRepository);
		
		SuiWindow window = createWindow(trainee, training, profession);
		
		window.display(player);
	}
	
	@NotNull
	private SuiWindow createWindow(Trainee trainee, Training training, Profession profession) {
		List<Skill> skills = new ArrayList<>(training.whatCanTraineeLearnRightNow(profession, trainee));
		SuiListBox listBox = new SuiListBox("Training", "Skills you can learn right now");
		
		addSkillsAsRows(trainee, training, skills, listBox);
		
		listBox.addOkButtonCallback("trainskill", (event, params) -> {
			int selectedRow = SuiListBox.getSelectedRow(params);
			
			if (isInvalidSelection(skills, selectedRow)) {
				return;
			}
			
			Skill skill = skills.get(selectedRow);
			
			training.trainSkill(trainee, skill);
		});
		
		return listBox;
	}
	
	private void addSkillsAsRows(Trainee trainee, Training training, List<Skill> skills, SuiListBox listBox) {
		for (Skill skill : skills) {
			String skillDisplayName = "@skl_n:" + skill.getKey();
			String requiredSkillPoints = getRequiredSkillPoints(training, skill, trainee);
			String requiredCredits = getRequiredCredits(training, trainee, skill);
			listBox.addListItem(skillDisplayName + " " + WHITE + "| " + requiredSkillPoints + " | " + requiredCredits);
		}
	}
	
	private String getRequiredSkillPoints(Training training, Skill skill, Trainee trainee) {
		int requiredSkillPoints = skill.getRequiredSkillPoints();
		boolean enoughAvailableSkillPoints = training.enoughAvailableSkillPoints(trainee, skill);
		
		String color;
		
		if (enoughAvailableSkillPoints) {
			color = GREEN;
		} else {
			color = RED;
		}
		
		return color + requiredSkillPoints + " skill points" + WHITE;
	}
	
	@NotNull
	private String getRequiredCredits(Training training, Trainee trainee, Skill skill) {
		String color;
		
		if (training.traineeHasEnoughCredits(trainee, skill)) {
			color = GREEN;
		} else {
			color = RED;
		}
		
		return color + skill.getRequiredCredits() + " credits" + WHITE;
	}
	
	private boolean isInvalidSelection(List<Skill> skills, int selectedRow) {
		return selectedRow < 0 || selectedRow >= skills.size();
	}
}

File: src/main/java/com/projectswg/holocore/services/gameplay/training/AvailableSkillsWindow.kt
/***********************************************************************************
 * Copyright (c) 2024 /// Project SWG /// www.projectswg.com                       *
 *                                                                                 *
 * ProjectSWG is the first NGE emulator for Star Wars Galaxies founded on          *
 * July 7th, 2011 after SOE announced the official shutdown of Star Wars Galaxies. *
 * Our goal is to create an emulator which will provide a server for players to    *
 * continue playing a game similar to the one they used to play. We are basing     *
 * it on the final publish of the game prior to end-game events.                   *
 *                                                                                 *
 * This file is part of Holocore.                                                  *
 *                                                                                 *
 * --------------------------------------------------------------------------------*
 *                                                                                 *
 * Holocore is free software: you can redistribute it and/or modify                *
 * it under the terms of the GNU Affero General Public License as                  *
 * published by the Free Software Foundation, either version 3 of the              *
 * License, or (at your option) any later version.                                 *
 *                                                                                 *
 * Holocore is distributed in the hope that it will be useful,                     *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of                  *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                   *
 * GNU Affero General Public License for more details.                             *
 *                                                                                 *
 * You should have received a copy of the GNU Affero General Public License        *
 * along with Holocore.  If not, see <http://www.gnu.org/licenses/>.               *
 ***********************************************************************************/
package com.projectswg.holocore.services.gameplay.training

import com.projectswg.common.data.sui.SuiEvent
import com.projectswg.holocore.resources.support.global.player.Player
import com.projectswg.holocore.resources.support.global.zone.sui.SuiListBox

class AvailableSkillsWindow {
	fun show(professionName: String?, player: Player) {
		val profession = Profession(professionName)
		val skillRepository = SkillsSdbSkillRepository()
		val trainee: Trainee = TraineeImpl(player, skillRepository)
		val training = Training(skillRepository)

		createWindow(trainee, training, profession, player)
	}

	private fun createWindow(trainee: Trainee, training: Training, profession: Profession, player: Player) {
		val skills: List<Skill> = ArrayList(training.whatCanTraineeLearnRightNow(profession, trainee))
		SuiListBox().run {
			title = "Training"
			prompt = "Skills you can learn right now"

			addSkillsAsRows(trainee, training, skills, this)

			addOkButtonCallback("trainskill") { _: SuiEvent, params: Map<String, String> ->
				val selectedRow = SuiListBox.getSelectedRow(params)
				if (isInvalidSelection(skills, selectedRow)) {
					return@addOkButtonCallback
				}

				val skill = skills[selectedRow]
				training.trainSkill(trainee, skill)
			}

			display(player)
		}
	}

	private fun addSkillsAsRows(trainee: Trainee, training: Training, skills: List<Skill>, listBox: SuiListBox) {
		for (skill in skills) {
			val skillDisplayName = "@skl_n:" + skill.key
			val requiredSkillPoints = getRequiredSkillPoints(training, skill, trainee)
			val requiredCredits = getRequiredCredits(training, trainee, skill)
			listBox.addListItem("$skillDisplayName $WHITE| $requiredSkillPoints | $requiredCredits")
		}
	}

	private fun getRequiredSkillPoints(training: Training, skill: Skill, trainee: Trainee): String {
		val requiredSkillPoints = skill.requiredSkillPoints
		val enoughAvailableSkillPoints = training.enoughAvailableSkillPoints(trainee, skill)

		val color = if (enoughAvailableSkillPoints) {
			GREEN
		} else {
			RED
		}

		return "$color$requiredSkillPoints skill points$WHITE"
	}

	private fun getRequiredCredits(training: Training, trainee: Trainee, skill: Skill): String {
		val color = if (training.traineeHasEnoughCredits(trainee, skill)) {
			GREEN
		} else {
			RED
		}

		return color + skill.requiredCredits + " credits" + WHITE
	}

	private fun isInvalidSelection(skills: List<Skill>, selectedRow: Int): Boolean {
		return selectedRow < 0 || selectedRow >= skills.size
	}

	companion object {
		private const val GREEN = "\\#00FF00"
		private const val RED = "\\#FF0000"
		private const val WHITE = "\\#FFFFFF"
	}
}
